purple-pill/js/controllers/fontexportController.js

```diff
@@ -32,140 +112,107 @@ function($scope, $http, sharedScope, $timeout) {
+              , glyphId = it.value['glyph_id']
+              ;
+            msg = totalInstances + " instances to export."
+            msg += "Calculating glyph '" + glyphId + "' (" + currentGlyph + " of " + totalGlyphs + ")"
+            msg += " of instance #" + (instanceIndex+1);
+            return msg;
+        }
+
+        function calculateGlyphsProgress(it, instanceIndex, totalInstances){
+            var percentage
+              , currentGlyph = it.value['current_glyph']
+              , totalGlyphs = it.value['total_glyphs']
+              ;
+            percentage = (instanceIndex + (currentGlyph / totalGlyphs)) / totalInstances;
+            return percentage;
         }
 
-        function exportFont_compute_glyphs(){
+        var index = 0
+          , totalInstances = exportObjects.length
+          , obj = exportObjects.pop()
+          ;
+        function exportFontComputeGlyphs(){
             var text
-              , current_glyph
-              , total_glyphs
-              , glyph_name
-              , it = generator.next()
+              , percentage
+              , zippedData
+              , it = obj.generator.next()
               ;
             if (!it.done){
-                current_glyph = it.value['current_glyph'];
-                total_glyphs = it.value['total_glyphs'];
-                glyph_id = it.value['glyph_id'];
-                text = total_instances + " instances to export. Calculating glyph '" + glyph_id + "' (" + (current_glyph+1) + " of " + total_glyphs + ") of instance #" + (current_instance+1);
-                setProgress(glyphs_phase_percentage * (current_instance + ((current_glyph+1) / total_glyphs))/total_instances, text);
-                $timeout(exportFont_compute_glyphs, UI_UPDATE_TIMESLICE);
+                text = exportingGlyphMessage(it, index);
+                percentage = calculateGlyphsProgress(it, index, totalInstances);
+                progress.set(percentage, text);
+                $timeout(exportFontComputeGlyphs, UI_UPDATE_TIMESLICE);
             } else {
-                delete generator;
-                if (++current_instance < total_instances) {
-                    generator = getExportGenerator(current_instance);
-                    $timeout(exportFont_compute_glyphs, UI_UPDATE_TIMESLICE);
+                index++;
+                delete obj.generator;
+
+                if (obj.fileFormat == "UFO"){
+                    zippedData = $scope.data.stateful.project.getZipFromIo(false, obj.io, obj.getFileName(), "uint8array");
+                    bundleFolder.file(obj.getFileName()+".zip", zippedData, {binary:true});
+                    delete zippedData;
+                } else { /* obj.fileFormat == "OTF" */
+                    var plainData = obj.io.readFile(false, obj.getFileName());
+                    bundleFolder.file(obj.getFileName(), plainData, {binary:true});
+                }
+                delete obj;
+
```

Comment:
I would put these checks first in the function. As they would prevent execution if there's nothing in `exportObjects` in the first place.